### PR TITLE
Add recipes for golang, Control and coconut

### PR DESCRIPTION
**Pull request by teo**
Contents:
  - `golang.sh` - replaces package `golang` from CentOS base repo if the latter is older than version `1.10.0`, this intentionally uses official binary tarballs from upstream (for both linux-amd64 and darwin) for a faster and more reliable build that doesn't rely on CentOS- or aliBuild-provided toolchains;
  - `coconut.sh` - the COntrol and CONfiguration UTility for the O² farm, builds with Go from repo AliceO2Group/Control (same as Control-OCCPlugin);
  - `control-core.sh` - main package for O² Control, builds with Go from repo AliceO2Group/Control (same as Control-OCCPlugin);
  - `control.sh` - metarecipe that pulls in as dependencies all of the above, plus Control-OCCPlugin which was previously merged;

\+ fixed workaround for `libgrpc++.so.1` symlink creation.

### CODE REVIEW

**Comment by dberzano** [comment]
@teo can't we have a toplevel one which uses them? Do you really need the three of them separately? To my understanding, RPM policies should be clarified at your meeting tomorrow, right? (_e.g._ having updatable PRMs, etc.)

**Comment by teo** [comment]
@dberzano the reasoning for 3 packages is as follows:
`Control-OCCPlugin`: C++ plugin, loaded throughout the O² farm by FairMQ processes. The same machines will generally also have the Control package installed, but this is a logically and technically separate component with a separate build system.
`Control`: the O² Control system, installed throughout the O² farm and other similar environments. Contains two executables, a core and an executor.
`coconut`: a command line utility which interacts with O² Control and Configuration. This one is not normally installed throughout the O² farm, but might be installed on some hosts in the farm and generally users' laptops. The full O² Control has no reason to be installed if the user only needs coconut to talk to an existing O² Control instance.

TL;DR: I'd rather not merge any of these, but if I really must, I'd merge `Control` with `Control-OCCPlugin`, and leave `coconut` separate.

**Comment by dberzano** [comment]
I am not suggesting to merge recipes, I am suggesting to create a meta-recipe called, I don't know, `Control-Tools`, which depends on the three/four recipes above. This makes _your_ life easier as you'll have to tell Jenkins to rebuild _that_ package too, not four different ones. And it also fixes potential problems with incompatible dependencies (you will have in any case to update the three/four packages altogether if you don't want to run into dependency problems).

File: golang.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+build_requires:
+  - curl
+prefer_system_check: |
+  which go && case `go version | sed -e 's/go version go//' | sed -e 's/ .*//'` in 0*|1.[0-9].*) exit 1 ;; esac
+incremental_recipe: |
+  case $ARCHITECTURE in
+    osx*) ARCH=darwin-amd64 ;;
+    *) ARCH=linux-amd64 ;;
+  esac
+  SOURCE=https://dl.google.com/go/go$PKGVERSION.$ARCH.tar.gz
```
Comment (sbinet) [inline golang.sh]: better use `https://golang.org/dl/go$PKGVERSION.$ARCH.tar.gz`

File: golang.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  mkdir -p $INSTALLROOT/etc/modulefiles && rsync -a --delete etc/modulefiles/ $INSTALLROOT/etc/modulefiles
+---
+#!/bin/bash -e
+
+case $ARCHITECTURE in
+  osx*) ARCH=darwin-amd64 ;;
+  *) ARCH=linux-amd64 ;;
+esac
+
+SOURCE=https://dl.google.com/go/go$PKGVERSION.$ARCH.tar.gz
```
Comment (sbinet) [inline golang.sh]: ditto :)

**Comment by ktf** [comment]
I would merge all of those, actually. I suspect that they build in a picosecond and that they are all a few megabytes each, why bothering optimising that? I think we have enough undocumented "bump this" "bump that".

File: CODEOWNERS [hunk 3]
```diff
@@ -2,11 +2,13 @@
 *                   @alisw/infrastructure-admins
 
 # Dataflow team
+coconut             @alisw/dataflow-experts
 common-o2*          @alisw/dataflow-experts
 configuration*      @alisw/dataflow-experts
 control*            @alisw/dataflow-experts
 datasampling*       @alisw/dataflow-experts
 flpproto*           @alisw/dataflow-experts
+golang              @alisw/dataflow-experts
```
Comment (dberzano) [inline CODEOWNERS]: Please assign the ownership of `golang` to `@alisw/infrastructure-admins`, thanks!

File: coconut.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  mkdir -p $INSTALLROOT/etc/modulefiles && rsync -a --delete etc/modulefiles/ $INSTALLROOT/etc/modulefiles
+---
+#!/bin/bash -e
+
+GOPATH=$PWD
+PATH=$GOPATH/bin:$PATH
+BUILD=$GOPATH/src/github.com/AliceO2Group/Control
+mkdir -p $BUILD
+rsync -a --delete $SOURCEDIR/ $BUILD/
+cd $BUILD
```
Comment (dberzano) [inline coconut.sh]: Use `pushd ... popd` (not `cd ... cd -`) with proper indentation, for instance:

```bash
pushd $MYDIR
  do stuff
  do more stuff
popd
```

File: coconut.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+make WHAT="coconut" all
+mkdir -p $INSTALLROOT/bin
+cp bin/* $INSTALLROOT/bin
```
Comment (dberzano) [inline coconut.sh]: Make sure you are retaining permissions. Cross-platform way of doing it is not to use `cp` but:

```bash
rsync -a --delete bin/ $INSTALLROOT/bin
```

(mind the final slashes, they matter a lot with `rsync`.)

File: coconut.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  global version
+  puts stderr "ALICE Modulefile for $PKGNAME $PKGVERSION-@@PKGREVISION@$PKGHASH@@"
+}
+set version $PKGVERSION-@@PKGREVISION@$PKGHASH@@
+module-whatis "ALICE Modulefile for $PKGNAME $PKGVERSION-@@PKGREVISION@$PKGHASH@@"
+# Dependencies
+module load BASE/1.0
+
+# Our environment
+set coconut_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
Comment (dberzano) [inline coconut.sh]: The convention (even if I see it has been violated in other places) is to have the package name UPPERCASE, so - even if this variable won't go in the user's environment - please use `COCONUT_ROOT`.

File: control-core.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  mkdir -p $INSTALLROOT/etc/modulefiles && rsync -a --delete etc/modulefiles/ $INSTALLROOT/etc/modulefiles
+---
+#!/bin/bash -e
+
+GOPATH=$PWD
+PATH=$GOPATH/bin:$PATH
+BUILD=$GOPATH/src/github.com/AliceO2Group/Control
+mkdir -p $BUILD
+rsync -a --delete $SOURCEDIR/ $BUILD/
+cd $BUILD
```
Comment (dberzano) [inline control-core.sh]: `pushd ... popd`

File: control-core.sh [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+make WHAT="octld octl-executor" all
+mkdir -p $INSTALLROOT/bin
+cp bin/* $INSTALLROOT/bin
```
Comment (dberzano) [inline control-core.sh]: `rsync` as said before

File: control-core.sh [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  puts stderr "ALICE Modulefile for $PKGNAME $PKGVERSION-@@PKGREVISION@$PKGHASH@@"
+}
+set version $PKGVERSION-@@PKGREVISION@$PKGHASH@@
+module-whatis "ALICE Modulefile for $PKGNAME $PKGVERSION-@@PKGREVISION@$PKGHASH@@"
+# Dependencies
+module load BASE/1.0 \\
+            ${PROTOBUF_VERSION:+protobuf/$PROTOBUF_VERSION-$PROTOBUF_REVISION}
+
+# Our environment
+set Control_Core_ROOT \$::env(BASEDIR)/$PKGNAME/\$version
```
Comment (dberzano) [inline control-core.sh]: `CONTROL_CORE_ROOT` uppercase

File: golang.sh [hunk 10]
```diff
@@ -0,0 +1,47 @@
+package: golang
+version: "1.10.2"
+build_requires:
+  - curl
+prefer_system_check: |
+  which go && case `go version | sed -e 's/go version go//' | sed -e 's/ .*//'` in 0*|1.[0-9].*) exit 1 ;; esac
```
Comment (dberzano) [inline golang.sh]: Do not use `which`, call `go version` directly instead (it returns `0` in case `go` exists). I know, this discussion was done already in other places: `which` might _(might)_ not be available on all platforms and in all cases it's better to invoke the target executable directly.

File: golang.sh [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+incremental_recipe: |
```
Comment (dberzano) [inline golang.sh]: This incremental recipe is useless, please drop it completely (it does precisely the same as the full one).

File: golang.sh [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    *) ARCH=linux-amd64 ;;
+  esac
+  SOURCE=https://golang.org/dl/go$PKGVERSION.$ARCH.tar.gz
+  curl -LO $SOURCE
+  tar --strip-components=1 -C $INSTALLROOT -xzf go$PKGVERSION.$ARCH.tar.gz
+  mkdir -p $INSTALLROOT/etc/modulefiles && rsync -a --delete etc/modulefiles/ $INSTALLROOT/etc/modulefiles
+---
+#!/bin/bash -e
+
+case $ARCHITECTURE in
```
Comment (dberzano) [inline golang.sh]: You can replace the full block with:

```bash
ARCH=$(uname|tr '[:upper:]' '[:lower:]')-amd64
```

File: coconut.sh [hunk 13]
```diff
@@ -0,0 +1,45 @@
+package: coconut
+version: "%(commit_hash)s"
+tag:  master
+build_requires:
+  - golang
+  - protobuf
+source: https://github.com/AliceO2Group/Control
+incremental_recipe: |
```
Comment (dberzano) [inline coconut.sh]: Incremental does the same as the base recipe, please drop the whole incremental recipe.

File: control-core.sh [hunk 14]
```diff
@@ -0,0 +1,46 @@
+package: Control-Core
+version: "%(commit_hash)s"
+tag:  master
+build_requires:
+  - golang
+  - protobuf
+source: https://github.com/AliceO2Group/Control
+incremental_recipe: |
```
Comment (dberzano) [inline control-core.sh]: Incremental recipe does the same as the base recipe (minus the `GOPATH` stuff, so how could it possibly work?), please drop it

**Comment by dberzano** [comment]
Just waiting for the tests to be green before merging (nothing should be impacted but it's better to check). After that, it can be merged.

**Comment by dberzano** [comment]
So: everything's fine but we need to make it work with the new split FairRoot (FairMQ notably). Waiting for it before unlocking.

**Comment by dberzano** [comment]
Let's wait for the tests now. Thanks!

**Comment by dberzano** [comment]
@teo I believe you'll be able to merge it yourself when tests are green.

**Comment by teo** [comment]
@dberzano I cannot even though I'm in alisw/dataflow-experts, I'm not allowed to approve my own PR.

**Comment by dberzano** [comment]
Hi @teo, @vascobarroso, let me clarify.

* You are a repo contributor but not an admin, therefore you _can_ merge _every_ pull request in alidist. With "merge" I mean: press the "Squash and merge" button
* Given that you are not an admin, however, you can do that if and only if:
  - Tests have passed,
  - Reviews from all the involved code owners have been received

GitHub is smart, so if you are in a certain mentioned group that's supposed to approve the PR but you are the author of the PR, you cannot review it. If you think of it, it does not make sense to review your own PR. So, you need _another_ person from your group to approve it (@vascobarroso in this case).
